### PR TITLE
Simulator Metrics

### PR DESCRIPTION
**Pull request by lawhead**
# Overview

Compile metrics summarizing all of the runs in a simulation.
* Outputs summary_data.json which collects the metrics from each session.json file in the run.
* Computes descriptive statistics for each field and outputs the results to the log files and console.
* Generates box plots for several statistics of interest.
* Outputs the mean duration of the runs.

## Ticket

https://www.pivotaltracker.com/story/show/188335185

## Contributions

- Also updated sim logging to use INFO, rather than DEBUG to clean up the log files.

## Test

- Unit tests
- Ran several simulations and inspected the output.

## Screenshots
Screenshot 2024-12-17 at 1 37 33 PM

Screenshot 2024-12-17 at 1 37 57 PM

### CODE REVIEW

File: bcipy/simulator/metrics.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+from bcipy.config import DEFAULT_ENCODING
+from bcipy.helpers.acquisition import max_inquiry_duration
+from bcipy.io.load import load_json_parameters
+from bcipy.io.save import save_json_data
+from bcipy.simulator.util.artifact import TOP_LEVEL_LOGGER_NAME
+
+logger = logging.getLogger(TOP_LEVEL_LOGGER_NAME)
+
+SUMMARY_DATA_FILE_NAME = "summary_data.json"
```
Comment (tab-cmd) [inline bcipy/simulator/metrics.py]: These could go in the top-level config.

File: bcipy/simulator/metrics.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    """List of paths to the session.json files"""
+    return [
+        Path(run_dir, "session.json")
+        for run_dir in sorted(Path(sim_dir).glob("run*/"))
+    ]
+
+
+def sim_parameters(sim_dir: str) -> Dict[str, Any]:
+    """Simulation parameters"""
+    return load_json_parameters(str(Path(sim_dir, "parameters.json")), True)
```
Comment (tab-cmd) [inline bcipy/simulator/metrics.py]: Pull default name from config? Return is `Parameters`

File: bcipy/simulator/util/artifact.py [hunk 3]
```diff
@@ -30,14 +30,14 @@ def configure_logger(log_path: str,
     """
```
Comment (tab-cmd) [inline bcipy/simulator/util/artifact.py:32]: It is not required for this PR, but we should talk about the name and configuration of this module. There are some top-level constants here, too. If we want to separate the simulator, we could implement a simulator/config.py.

File: bcipy/simulator/metrics.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    """Add all items of interest from the session data."""
+    keys = [
+        'total_number_series', 'total_inquiries', 'total_selections',
+        'inquiries_per_selection'
+    ]
+    for key in keys:
+        add_item(combined_metrics, key, session_dict[key])
+
+    for key in session_dict['task_summary'].keys():
+        if 'switch' not in key and 'rate' not in key:
```
Comment (tab-cmd) [inline bcipy/simulator/metrics.py]: some context here would be helpful

File: bcipy/simulator/metrics.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    ]
+
+
+def sim_parameters(sim_dir: str) -> Dict[str, Any]:
+    """Simulation parameters"""
+    return load_json_parameters(str(Path(sim_dir, "parameters.json")), True)
+
+
+def summarize(sim_dir: str) -> Dict[str, List[Any]]:
+    """Summarize all session runs."""
```
Comment (tab-cmd) [inline bcipy/simulator/metrics.py]: We should document what we expect the data/dir to look like

File: bcipy/simulator/metrics.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    add_item(
+        combined_metrics, 'total_seconds',
+        calculate_duration(session_dict['total_inquiries'], inquiry_seconds))
+
+
+def session_paths(sim_dir: str) -> List[Path]:
+    """List of paths to the session.json files"""
+    return [
+        Path(run_dir, "session.json")
+        for run_dir in sorted(Path(sim_dir).glob("run*/"))
```
Comment (tab-cmd) [inline bcipy/simulator/metrics.py]: Is this a dir name we can pull from elsewhere?

File: bcipy/simulator/metrics.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+def add_item(container: Dict[str, List], key: str, value: Any):
+    """Add or append value"""
+    if key in container:
+        container[key].append(value)
+    else:
+        container[key] = [value]
+
+
+def get_final_typed(session_dict: Dict) -> str:
```
Comment (tab-cmd) [inline bcipy/simulator/metrics.py]: What do you expect session dict to look like?

File: bcipy/simulator/metrics.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
+"""Summarize metrics across simulator runs."""
```
Comment (tab-cmd) [inline bcipy/simulator/metrics.py]: This could go in data or util?

**Review by tab-cmd — APPROVED** [review]
Great addition! The README needs to be updated a bit to reflect the change in expected output. 

Suggestions:
What do you think about having a run-name argument? We could use it in the sim dir name and/or metric png title to help with multiple comparisons. 

When outputting the command generated from the GUI, we should add quotations around string paths to prevent issues with spaces in dir names.

File: bcipy/simulator/metrics.py (hunk 6 above)
Comment (lawhead) [reply bcipy/simulator/metrics.py]: Yes, I can add that to the artifact module.

File: bcipy/simulator/util/artifact.py (hunk 3 above)
Comment (lawhead) [reply bcipy/simulator/util/artifact.py:32]: Sounds good. Let's discuss.

**Pull request merged**